File: jscomp/stdlib/bigarray.ml
(***********************************************************************)
(*                                                                     *)
(*                                OCaml                                *)
(*                                                                     *)
(*          Manuel Serrano et Xavier Leroy, INRIA Rocquencourt         *)
(*                                                                     *)
(*  Copyright 2000 Institut National de Recherche en Informatique et   *)
(*  en Automatique.  All rights reserved.  This file is distributed    *)
(*  under the terms of the GNU Library General Public License, with    *)
(*  the special exception on linking described in file ../../LICENSE.  *)
(*                                                                     *)
(***********************************************************************)

(* Module [Bigarray]: large, multi-dimensional, numerical arrays *)

external init : unit -> unit = "caml_ba_init"

let _ = init()

type float32_elt = Float32_elt
type float64_elt = Float64_elt
type int8_signed_elt = Int8_signed_elt
type int8_unsigned_elt = Int8_unsigned_elt
type int16_signed_elt = Int16_signed_elt
type int16_unsigned_elt = Int16_unsigned_elt
type int32_elt = Int32_elt
type int64_elt = Int64_elt
type int_elt = Int_elt
type nativeint_elt = Nativeint_elt
type complex32_elt = Complex32_elt
type complex64_elt = Complex64_elt

type ('a, 'b) kind =
    Float32 : (float, float32_elt) kind
  | Float64 : (float, float64_elt) kind
  | Int8_signed : (int, int8_signed_elt) kind
  | Int8_unsigned : (int, int8_unsigned_elt) kind
  | Int16_signed : (int, int16_signed_elt) kind
  | Int16_unsigned : (int, int16_unsigned_elt) kind
  | Int32 : (int32, int32_elt) kind
  | Int64 : (int64, int64_elt) kind
  | Int : (int, int_elt) kind
  | Nativeint : (nativeint, nativeint_elt) kind
  | Complex32 : (Complex.t, complex32_elt) kind
  | Complex64 : (Complex.t, complex64_elt) kind
  | Char : (char, int8_unsigned_elt) kind

(* Keep those constants in sync with the caml_ba_kind enumeration
   in bigarray.h *)

let float32 = Float32
let float64 = Float64
let int8_signed = Int8_signed
let int8_unsigned = Int8_unsigned
let int16_signed = Int16_signed
let int16_unsigned = Int16_unsigned
let int32 = Int32
let int64 = Int64
let int = Int
let nativeint = Nativeint
let complex32 = Complex32
let complex64 = Complex64
let char = Char

type c_layout = C_layout_typ
type fortran_layout = Fortran_layout_typ

type 'a layout =
    C_layout: c_layout layout
  | Fortran_layout: fortran_layout layout

(* Keep those constants in sync with the caml_ba_layout enumeration
   in bigarray.h *)

let c_layout = C_layout
let fortran_layout = Fortran_layout

module Genarray = struct
  type ('a, 'b, 'c) t
  external create: ('a, 'b) kind -> 'c layout -> int array -> ('a, 'b, 'c) t
     = "caml_ba_create"
  external get: ('a, 'b, 'c) t -> int array -> 'a
     = "caml_ba_get_generic"
  external set: ('a, 'b, 'c) t -> int array -> 'a -> unit
     = "caml_ba_set_generic"
  external num_dims: ('a, 'b, 'c) t -> int = "caml_ba_num_dims"
  external nth_dim: ('a, 'b, 'c) t -> int -> int = "caml_ba_dim"
  let dims a =
    let n = num_dims a in
    let d = Array.make n 0 in
    for i = 0 to n-1 do d.(i) <- nth_dim a i done;
    d
  external kind: ('a, 'b, 'c) t -> ('a, 'b) kind = "caml_ba_kind"
  external layout: ('a, 'b, 'c) t -> 'c layout = "caml_ba_layout"

  external sub_left: ('a, 'b, c_layout) t -> int -> int -> ('a, 'b, c_layout) t
     = "caml_ba_sub"
  external sub_right: ('a, 'b, fortran_layout) t -> int -> int ->
                          ('a, 'b, fortran_layout) t
     = "caml_ba_sub"
  external slice_left: ('a, 'b, c_layout) t -> int array ->
                          ('a, 'b, c_layout) t
     = "caml_ba_slice"
  external slice_right: ('a, 'b, fortran_layout) t -> int array ->
                          ('a, 'b, fortran_layout) t
     = "caml_ba_slice"
  external blit: ('a, 'b, 'c) t -> ('a, 'b, 'c) t -> unit
     = "caml_ba_blit"
  external fill: ('a, 'b, 'c) t -> 'a -> unit = "caml_ba_fill"
  external map_internal: Unix.file_descr -> ('a, 'b) kind -> 'c layout ->
                     bool -> int array -> int64 -> ('a, 'b, 'c) t
     = "caml_ba_map_file_bytecode" "caml_ba_map_file"
  let map_file fd ?(pos = 0L) kind layout shared dims =
    map_internal fd kind layout shared dims pos
end

module Array1 = struct
  type ('a, 'b, 'c) t = ('a, 'b, 'c) Genarray.t
  let create kind layout dim =
    Genarray.create kind layout [|dim|]
  external get: ('a, 'b, 'c) t -> int -> 'a = "%caml_ba_ref_1"
  external set: ('a, 'b, 'c) t -> int -> 'a -> unit = "%caml_ba_set_1"
  external unsafe_get: ('a, 'b, 'c) t -> int -> 'a = "%caml_ba_unsafe_ref_1"
  external unsafe_set: ('a, 'b, 'c) t -> int -> 'a -> unit
     = "%caml_ba_unsafe_set_1"
  external dim: ('a, 'b, 'c) t -> int = "%caml_ba_dim_1"
  external kind: ('a, 'b, 'c) t -> ('a, 'b) kind = "caml_ba_kind"
  external layout: ('a, 'b, 'c) t -> 'c layout = "caml_ba_layout"
  external sub: ('a, 'b, 'c) t -> int -> int -> ('a, 'b, 'c) t = "caml_ba_sub"
  external blit: ('a, 'b, 'c) t -> ('a, 'b, 'c) t -> unit = "caml_ba_blit"
  external fill: ('a, 'b, 'c) t -> 'a -> unit = "caml_ba_fill"
  let of_array (type t) kind (layout: t layout) data =
    let ba = create kind layout (Array.length data) in
    let ofs =
      match layout with
        C_layout -> 0
      | Fortran_layout -> 1
    in
    for i = 0 to Array.length data - 1 do unsafe_set ba (i + ofs) data.(i) done;
    ba
  let map_file fd ?pos kind layout shared dim =
    Genarray.map_file fd ?pos kind layout shared [|dim|]
end

module Array2 = struct
  type ('a, 'b, 'c) t = ('a, 'b, 'c) Genarray.t
  let create kind layout dim1 dim2 =
    Genarray.create kind layout [|dim1; dim2|]
  external get: ('a, 'b, 'c) t -> int -> int -> 'a = "%caml_ba_ref_2"
  external set: ('a, 'b, 'c) t -> int -> int -> 'a -> unit = "%caml_ba_set_2"
  external unsafe_get: ('a, 'b, 'c) t -> int -> int -> 'a
     = "%caml_ba_unsafe_ref_2"
  external unsafe_set: ('a, 'b, 'c) t -> int -> int -> 'a -> unit
     = "%caml_ba_unsafe_set_2"
  external dim1: ('a, 'b, 'c) t -> int = "%caml_ba_dim_1"
  external dim2: ('a, 'b, 'c) t -> int = "%caml_ba_dim_2"
  external kind: ('a, 'b, 'c) t -> ('a, 'b) kind = "caml_ba_kind"
  external layout: ('a, 'b, 'c) t -> 'c layout = "caml_ba_layout"
  external sub_left: ('a, 'b, c_layout) t -> int -> int -> ('a, 'b, c_layout) t
     = "caml_ba_sub"
  external sub_right:
    ('a, 'b, fortran_layout) t -> int -> int -> ('a, 'b, fortran_layout) t
     = "caml_ba_sub"
  let slice_left a n = Genarray.slice_left a [|n|]
  let slice_right a n = Genarray.slice_right a [|n|]
  external blit: ('a, 'b, 'c) t -> ('a, 'b, 'c) t -> unit = "caml_ba_blit"
  external fill: ('a, 'b, 'c) t -> 'a -> unit = "caml_ba_fill"
  let of_array (type t) kind (layout: t layout) data =
    let dim1 = Array.length data in
    let dim2 = if dim1 = 0 then 0 else Array.length data.(0) in
    let ba = create kind layout dim1 dim2 in
    let ofs =
      match layout with
        C_layout -> 0
      | Fortran_layout -> 1
    in
    for i = 0 to dim1 - 1 do
      let row = data.(i) in
      if Array.length row <> dim2 then
        invalid_arg("Bigarray.Array2.of_array: non-rectangular data");
      for j = 0 to dim2 - 1 do
        unsafe_set ba (i + ofs) (j + ofs) row.(j)
      done
    done;
    ba
  let map_file fd ?pos kind layout shared dim1 dim2 =
    Genarray.map_file fd ?pos kind layout shared [|dim1;dim2|]
end

module Array3 = struct
  type ('a, 'b, 'c) t = ('a, 'b, 'c) Genarray.t
  let create kind layout dim1 dim2 dim3 =
    Genarray.create kind layout [|dim1; dim2; dim3|]
  external get: ('a, 'b, 'c) t -> int -> int -> int -> 'a = "%caml_ba_ref_3"
  external set: ('a, 'b, 'c) t -> int -> int -> int -> 'a -> unit
     = "%caml_ba_set_3"
  external unsafe_get: ('a, 'b, 'c) t -> int -> int -> int -> 'a
     = "%caml_ba_unsafe_ref_3"
  external unsafe_set: ('a, 'b, 'c) t -> int -> int -> int -> 'a -> unit
     = "%caml_ba_unsafe_set_3"
  external dim1: ('a, 'b, 'c) t -> int = "%caml_ba_dim_1"
  external dim2: ('a, 'b, 'c) t -> int = "%caml_ba_dim_2"
  external dim3: ('a, 'b, 'c) t -> int = "%caml_ba_dim_3"
  external kind: ('a, 'b, 'c) t -> ('a, 'b) kind = "caml_ba_kind"
  external layout: ('a, 'b, 'c) t -> 'c layout = "caml_ba_layout"
  external sub_left: ('a, 'b, c_layout) t -> int -> int -> ('a, 'b, c_layout) t
     = "caml_ba_sub"
  external sub_right:
     ('a, 'b, fortran_layout) t -> int -> int -> ('a, 'b, fortran_layout) t
     = "caml_ba_sub"
  let slice_left_1 a n m = Genarray.slice_left a [|n; m|]
  let slice_right_1 a n m = Genarray.slice_right a [|n; m|]
  let slice_left_2 a n = Genarray.slice_left a [|n|]
  let slice_right_2 a n = Genarray.slice_right a [|n|]
  external blit: ('a, 'b, 'c) t -> ('a, 'b, 'c) t -> unit = "caml_ba_blit"
  external fill: ('a, 'b, 'c) t -> 'a -> unit = "caml_ba_fill"
  let of_array (type t) kind (layout: t layout) data =
    let dim1 = Array.length data in
    let dim2 = if dim1 = 0 then 0 else Array.length data.(0) in
    let dim3 = if dim2 = 0 then 0 else Array.length data.(0).(0) in
    let ba = create kind layout dim1 dim2 dim3 in
    let ofs =
      match layout with
        C_layout -> 0
      | Fortran_layout -> 1
    in
    for i = 0 to dim1 - 1 do
      let row = data.(i) in
      if Array.length row <> dim2 then
        invalid_arg("Bigarray.Array3.of_array: non-cubic data");
      for j = 0 to dim2 - 1 do
        let col = row.(j) in
        if Array.length col <> dim3 then
          invalid_arg("Bigarray.Array3.of_array: non-cubic data");
        for k = 0 to dim3 - 1 do
          unsafe_set ba (i + ofs) (j + ofs) (k + ofs) col.(k)
        done
      done
    done;
    ba
  let map_file fd ?pos kind layout shared dim1 dim2 dim3 =
    Genarray.map_file fd ?pos kind layout shared [|dim1;dim2;dim3|]
end

external genarray_of_array1: ('a, 'b, 'c) Array1.t -> ('a, 'b, 'c) Genarray.t
   = "%identity"
external genarray_of_array2: ('a, 'b, 'c) Array2.t -> ('a, 'b, 'c) Genarray.t
   = "%identity"
external genarray_of_array3: ('a, 'b, 'c) Array3.t -> ('a, 'b, 'c) Genarray.t
   = "%identity"
let array1_of_genarray a =
  if Genarray.num_dims a = 1 then a
  else invalid_arg "Bigarray.array1_of_genarray"
let array2_of_genarray a =
  if Genarray.num_dims a = 2 then a
  else invalid_arg "Bigarray.array2_of_genarray"
let array3_of_genarray a =
  if Genarray.num_dims a = 3 then a
  else invalid_arg "Bigarray.array3_of_genarray"

external reshape:
   ('a, 'b, 'c) Genarray.t -> int array -> ('a, 'b, 'c) Genarray.t
   = "caml_ba_reshape"
let reshape_1 a dim1 = reshape a [|dim1|]
let reshape_2 a dim1 dim2 = reshape a [|dim1;dim2|]
let reshape_3 a dim1 dim2 dim3 = reshape a [|dim1;dim2;dim3|]

(* Force caml_ba_get_{1,2,3,N} to be linked in, since we don't refer
   to those primitives directly in this file *)

let _ =
  let _ = Genarray.get in
  let _ = Array1.get in
  let _ = Array2.get in
  let _ = Array3.get in
  ()

external get1: unit -> unit = "caml_ba_get_1"
external get2: unit -> unit = "caml_ba_get_2"
external get3: unit -> unit = "caml_ba_get_3"
external set1: unit -> unit = "caml_ba_set_1"
external set2: unit -> unit = "caml_ba_set_2"
external set3: unit -> unit = "caml_ba_set_3"

File: lib/es6/bigarray.js


import * as Sys from "./sys.js";
import * as $$Array from "./array.js";
import * as Caml_array from "./caml_array.js";
import * as Caml_int32 from "./caml_int32.js";
import * as Caml_int64 from "./caml_int64.js";
import * as Pervasives from "./pervasives.js";
import * as Caml_external_polyfill from "./caml_external_polyfill.js";

function kind_size_in_bytes(param) {
  switch (param) {
    case /* Int16_signed */4 :
    case /* Int16_unsigned */5 :
        return 2;
    case /* Float32 */0 :
    case /* Int32 */6 :
        return 4;
    case /* Int */8 :
    case /* Nativeint */9 :
        return Sys.word_size / 8 | 0;
    case /* Float64 */1 :
    case /* Int64 */7 :
    case /* Complex32 */10 :
        return 8;
    case /* Complex64 */11 :
        return 16;
    case /* Int8_signed */2 :
    case /* Int8_unsigned */3 :
    case /* Char */12 :
        return 1;
    
  }
}

function dims(a) {
  var n = Caml_external_polyfill.resolve("caml_ba_num_dims")(a);
  var d = Caml_array.caml_make_vect(n, 0);
  for(var i = 0; i < n; ++i){
    Caml_array.caml_array_set(d, i, Caml_external_polyfill.resolve("caml_ba_dim")(a, i));
  }
  return d;
}

function size_in_bytes(arr) {
  return Caml_int32.imul(kind_size_in_bytes(Caml_external_polyfill.resolve("caml_ba_kind")(arr)), $$Array.fold_left(Caml_int32.imul, 1, dims(arr)));
}

function map_file(fd, posOpt, kind, layout, shared, dims) {
  var pos = posOpt !== undefined ? posOpt : Caml_int64.zero;
  return Caml_external_polyfill.resolve("caml_ba_map_file_bytecode")(fd, kind, layout, shared, dims, pos);
}

var Genarray = {
  dims: dims,
  size_in_bytes: size_in_bytes,
  map_file: map_file
};

function create(kind, layout) {
  return Caml_external_polyfill.resolve("caml_ba_create")(kind, layout, []);
}

function get(arr) {
  return Caml_external_polyfill.resolve("caml_ba_get_generic")(arr, []);
}

function set(arr) {
  var partial_arg = [];
  return function (param) {
    return Caml_external_polyfill.resolve("caml_ba_set_generic")(arr, partial_arg, param);
  };
}

function size_in_bytes$1(arr) {
  return kind_size_in_bytes(Caml_external_polyfill.resolve("caml_ba_kind")(arr));
}

function of_value(kind, layout, v) {
  var a = create(kind, layout);
  set(a)(v);
  return a;
}

function create$1(kind, layout, dim) {
  return Caml_external_polyfill.resolve("caml_ba_create")(kind, layout, [dim]);
}

function size_in_bytes$2(arr) {
  return Caml_int32.imul(kind_size_in_bytes(Caml_external_polyfill.resolve("caml_ba_kind")(arr)), Caml_external_polyfill.resolve("caml_ba_dim_1")(arr));
}

function slice(a, n) {
  Caml_external_polyfill.resolve("caml_ba_layout")(a);
  return Caml_external_polyfill.resolve("caml_ba_slice")(a, [n]);
}

function of_array(kind, layout, data) {
  var ba = create$1(kind, layout, data.length);
  var ofs = layout ? 1 : 0;
  for(var i = 0 ,i_finish = data.length; i < i_finish; ++i){
    Caml_external_polyfill.resolve("caml_ba_set_1")(ba, i + ofs | 0, Caml_array.caml_array_get(data, i));
  }
  return ba;
}

function map_file$1(fd, pos, kind, layout, shared, dim) {
  return map_file(fd, pos, kind, layout, shared, [dim]);
}

function create$2(kind, layout, dim1, dim2) {
  return Caml_external_polyfill.resolve("caml_ba_create")(kind, layout, [
              dim1,
              dim2
            ]);
}

function size_in_bytes$3(arr) {
  return Caml_int32.imul(Caml_int32.imul(kind_size_in_bytes(Caml_external_polyfill.resolve("caml_ba_kind")(arr)), Caml_external_polyfill.resolve("caml_ba_dim_1")(arr)), Caml_external_polyfill.resolve("caml_ba_dim_2")(arr));
}

function slice_left(a, n) {
  return Caml_external_polyfill.resolve("caml_ba_slice")(a, [n]);
}

function slice_right(a, n) {
  return Caml_external_polyfill.resolve("caml_ba_slice")(a, [n]);
}

function of_array$1(kind, layout, data) {
  var dim1 = data.length;
  var dim2 = dim1 === 0 ? 0 : Caml_array.caml_array_get(data, 0).length;
  var ba = create$2(kind, layout, dim1, dim2);
  var ofs = layout ? 1 : 0;
  for(var i = 0; i < dim1; ++i){
    var row = Caml_array.caml_array_get(data, i);
    if (row.length !== dim2) {
      Pervasives.invalid_arg("Bigarray.Array2.of_array: non-rectangular data");
    }
    for(var j = 0; j < dim2; ++j){
      Caml_external_polyfill.resolve("caml_ba_set_2")(ba, i + ofs | 0, j + ofs | 0, Caml_array.caml_array_get(row, j));
    }
  }
  return ba;
}

function map_file$2(fd, pos, kind, layout, shared, dim1, dim2) {
  return map_file(fd, pos, kind, layout, shared, [
              dim1,
              dim2
            ]);
}

function create$3(kind, layout, dim1, dim2, dim3) {
  return Caml_external_polyfill.resolve("caml_ba_create")(kind, layout, [
              dim1,
              dim2,
              dim3
            ]);
}

function size_in_bytes$4(arr) {
  return Caml_int32.imul(Caml_int32.imul(Caml_int32.imul(kind_size_in_bytes(Caml_external_polyfill.resolve("caml_ba_kind")(arr)), Caml_external_polyfill.resolve("caml_ba_dim_1")(arr)), Caml_external_polyfill.resolve("caml_ba_dim_2")(arr)), Caml_external_polyfill.resolve("caml_ba_dim_3")(arr));
}

function slice_left_1(a, n, m) {
  return Caml_external_polyfill.resolve("caml_ba_slice")(a, [
              n,
              m
            ]);
}

function slice_right_1(a, n, m) {
  return Caml_external_polyfill.resolve("caml_ba_slice")(a, [
              n,
              m
            ]);
}

function slice_left_2(a, n) {
  return Caml_external_polyfill.resolve("caml_ba_slice")(a, [n]);
}

function slice_right_2(a, n) {
  return Caml_external_polyfill.resolve("caml_ba_slice")(a, [n]);
}

function of_array$2(kind, layout, data) {
  var dim1 = data.length;
  var dim2 = dim1 === 0 ? 0 : Caml_array.caml_array_get(data, 0).length;
  var dim3 = dim2 === 0 ? 0 : Caml_array.caml_array_get(Caml_array.caml_array_get(data, 0), 0).length;
  var ba = create$3(kind, layout, dim1, dim2, dim3);
  var ofs = layout ? 1 : 0;
  for(var i = 0; i < dim1; ++i){
    var row = Caml_array.caml_array_get(data, i);
    if (row.length !== dim2) {
      Pervasives.invalid_arg("Bigarray.Array3.of_array: non-cubic data");
    }
    for(var j = 0; j < dim2; ++j){
      var col = Caml_array.caml_array_get(row, j);
      if (col.length !== dim3) {
        Pervasives.invalid_arg("Bigarray.Array3.of_array: non-cubic data");
      }
      for(var k = 0; k < dim3; ++k){
        Caml_external_polyfill.resolve("caml_ba_set_3")(ba, i + ofs | 0, j + ofs | 0, k + ofs | 0, Caml_array.caml_array_get(col, k));
      }
    }
  }
  return ba;
}

function map_file$3(fd, pos, kind, layout, shared, dim1, dim2, dim3) {
  return map_file(fd, pos, kind, layout, shared, [
              dim1,
              dim2,
              dim3
            ]);
}

function array0_of_genarray(a) {
  if (Caml_external_polyfill.resolve("caml_ba_num_dims")(a) === 0) {
    return a;
  } else {
    return Pervasives.invalid_arg("Bigarray.array0_of_genarray");
  }
}

function array1_of_genarray(a) {
  if (Caml_external_polyfill.resolve("caml_ba_num_dims")(a) === 1) {
    return a;
  } else {
    return Pervasives.invalid_arg("Bigarray.array1_of_genarray");
  }
}

function array2_of_genarray(a) {
  if (Caml_external_polyfill.resolve("caml_ba_num_dims")(a) === 2) {
    return a;
  } else {
    return Pervasives.invalid_arg("Bigarray.array2_of_genarray");
  }
}

function array3_of_genarray(a) {
  if (Caml_external_polyfill.resolve("caml_ba_num_dims")(a) === 3) {
    return a;
  } else {
    return Pervasives.invalid_arg("Bigarray.array3_of_genarray");
  }
}

function reshape_0(a) {
  return Caml_external_polyfill.resolve("caml_ba_reshape")(a, []);
}

function reshape_1(a, dim1) {
  return Caml_external_polyfill.resolve("caml_ba_reshape")(a, [dim1]);
}

function reshape_2(a, dim1, dim2) {
  return Caml_external_polyfill.resolve("caml_ba_reshape")(a, [
              dim1,
              dim2
            ]);
}

function reshape_3(a, dim1, dim2, dim3) {
  return Caml_external_polyfill.resolve("caml_ba_reshape")(a, [
              dim1,
              dim2,
              dim3
            ]);
}

var float32 = /* Float32 */0;

var float64 = /* Float64 */1;

var complex32 = /* Complex32 */10;

var complex64 = /* Complex64 */11;

var int8_signed = /* Int8_signed */2;

var int8_unsigned = /* Int8_unsigned */3;

var int16_signed = /* Int16_signed */4;

var int16_unsigned = /* Int16_unsigned */5;

var $$int = /* Int */8;

var int32 = /* Int32 */6;

var int64 = /* Int64 */7;

var nativeint = /* Nativeint */9;

var $$char = /* Char */12;

var c_layout = /* C_layout */0;

var fortran_layout = /* Fortran_layout */1;

function Array0_change_layout(prim, prim$1) {
  return Caml_external_polyfill.resolve("caml_ba_change_layout")(prim, prim$1);
}

var Array0 = {
  create: create,
  change_layout: Array0_change_layout,
  size_in_bytes: size_in_bytes$1,
  get: get,
  set: set,
  of_value: of_value
};

function Array1_change_layout(prim, prim$1) {
  return Caml_external_polyfill.resolve("caml_ba_change_layout")(prim, prim$1);
}

var Array1 = {
  create: create$1,
  change_layout: Array1_change_layout,
  size_in_bytes: size_in_bytes$2,
  slice: slice,
  of_array: of_array,
  map_file: map_file$1
};

function Array2_change_layout(prim, prim$1) {
  return Caml_external_polyfill.resolve("caml_ba_change_layout")(prim, prim$1);
}

var Array2 = {
  create: create$2,
  change_layout: Array2_change_layout,
  size_in_bytes: size_in_bytes$3,
  slice_left: slice_left,
  slice_right: slice_right,
  of_array: of_array$1,
  map_file: map_file$2
};

function Array3_change_layout(prim, prim$1) {
  return Caml_external_polyfill.resolve("caml_ba_change_layout")(prim, prim$1);
}

var Array3 = {
  create: create$3,
  change_layout: Array3_change_layout,
  size_in_bytes: size_in_bytes$4,
  slice_left_1: slice_left_1,
  slice_right_1: slice_right_1,
  slice_left_2: slice_left_2,
  slice_right_2: slice_right_2,
  of_array: of_array$2,
  map_file: map_file$3
};

function reshape(prim, prim$1) {
  return Caml_external_polyfill.resolve("caml_ba_reshape")(prim, prim$1);
}

export {
  float32 ,
  float64 ,
  complex32 ,
  complex64 ,
  int8_signed ,
  int8_unsigned ,
  int16_signed ,
  int16_unsigned ,
  $$int ,
  int32 ,
  int64 ,
  nativeint ,
  $$char ,
  kind_size_in_bytes ,
  c_layout ,
  fortran_layout ,
  Genarray ,
  Array0 ,
  Array1 ,
  Array2 ,
  Array3 ,
  array0_of_genarray ,
  array1_of_genarray ,
  array2_of_genarray ,
  array3_of_genarray ,
  reshape ,
  reshape_0 ,
  reshape_1 ,
  reshape_2 ,
  reshape_3 ,
  
}
/* No side effect */
